circuits/cpp/.clang-tidy

```diff
@@ -12,72 +12,147 @@
+    performance-for-range-copy
+    performance-implicit-conversion-in-loop
+    performance-inefficient-algorithm
+    performance-inefficient-string-concatenation
+    performance-inefficient-vector-operation
+    performance-move-const-arg
+    performance-move-constructor-init
+    performance-no-automatic-move
+    performance-no-int-to-ptr
+    performance-noexcept-move-constructor
+    performance-trivially-destructible
+    performance-type-promotion-in-math-fn
+    performance-unnecessary-copy-initialization
+    portability-restrict-system-includes
+    portability-simd-intrinsics
+    portability-std-allocator-const
+    -bugprone-unchecked-optional-access
+    -bugprone-unhandled-self-assignment
+    -clang-analyzer-core.UndefinedBinaryOperatorResult
+    -clang-analyzer-cplusplus.NewDeleteLeaks
+    -clang-analyzer-optin.performance.Padding
+    -cert-err58-cpp
+    -cert-oop54-cpp
+    -cppcoreguidelines-avoid-non-const-global-variables
+    -cppcoreguidelines-avoid-magic-numbers
+    -cppcoreguidelines-c-copy-assignment-signature
+    -cppcoreguidelines-no-malloc
+    -cppcoreguidelines-owning-memory
+    -cppcoreguidelines-pro-type-cstyle-cast
+    -cppcoreguidelines-pro-type-reinterpret-cast
+    -cppcoreguidelines-special-member-functions
+    -google-build-using-namespace
+    -google-global-names-in-headers
+    -google-readability-casting
+    -misc-definitions-in-headers
+    -misc-no-recursion
+    -misc-unconventional-assign-operator
+    -modernize-return-braced-init-list
+    -performance-unnecessary-value-param
+    -readability-function-cognitive-complexity
+    -readability-magic-numbers
 '
 
-# We treat all warnings as errors except for these few.
-# Some of these exceptions like 'google-build-using-namespace'
-# we should be able to manually fix project-wide and then
-# remove from the omissions list.
-WarningsAsErrors: '
-    *,
-    -bugprone-unchecked-optional-access,
-    -bugprone-unhandled-self-assignment,
-    -clang-analyzer-core.UndefinedBinaryOperatorResult,
-    -clang-analyzer-cplusplus.NewDeleteLeaks,
-    -clang-analyzer-optin.performance.Padding,
-    -cert-err58-cpp,
-    -cert-oop54-cpp,
-    -cppcoreguidelines-avoid-non-const-global-variables,
-    -cppcoreguidelines-avoid-magic-numbers,
-    -cppcoreguidelines-c-copy-assignment-signature,
-    -cppcoreguidelines-no-malloc,
-    -cppcoreguidelines-owning-memory,
-    -cppcoreguidelines-pro-type-cstyle-cast,
-    -cppcoreguidelines-pro-type-reinterpret-cast,
-    -cppcoreguidelines-special-member-functions,
-    -google-build-using-namespace,
-    -google-global-names-in-headers,
-    -google-readability-casting,
-    -misc-definitions-in-headers,
-    -misc-no-recursion,
-    -misc-unconventional-assign-operator,
-    -modernize-return-braced-init-list,
-    -performance-unnecessary-value-param,
-    -readability-function-cognitive-complexity,
-    -readability-magic-numbers,
-'
+# We treat all warnings as errors.
+WarningsAsErrors: '*'
```

Comment:
IIRC these exceptions were necessary because we aren't actually following all of these rules and some of the rules are buggy. I'm surprised you were able to run `clang-tidy` with _all_ warnings as errors without many failures.

Reply:
This is a subset. I hope once we look into it a bit more we can revert the old ones

Reply:
I'm confused.... Doesn't your change above make ALL warnings become errors?

Reply:
And my original code chose specific rules to NOT turn into errors.

Reply:
Right but the list above it is the thing. Those rules that are not errors are not part of this abbreviated list at all

Reply:
"We treat all warnings as errors" is true but at the same time, what we select as warnings is very explicit and slimmed down (does not include any of the things that were warnings only before)

Reply:
Ah, understood.